src/main/java/com/icthh/xm/uaa/service/query/UserQueryService.java

```diff
@@ -32,31 +38,126 @@
+            ofNullable(filterQuery.getActivated()).map(fn -> buildSpecification(fn, User_.activated))
         );
     }
 
-    private Specification<User> getLoginSpecification(StringFilter loginFilter) {
-        Function<Root<User>, Join<User, UserLogin>> functionToEntity = userRoot -> (Join<User, UserLogin>)userRoot.fetch(User_.logins);
+    private Specification<User> getLoginSpecificationForStrict(StringFilter loginFilter) {
+        Function<Root<User>, Join<User, UserLogin>> functionToEntity = userRoot -> (Join<User, UserLogin>) userRoot.fetch(User_.logins);
         Function<Join<User, UserLogin>, Expression<String>> entityToColumn = entity -> entity.get(UserLogin_.login);
         return buildSpecification(loginFilter, functionToEntity.andThen(entityToColumn));
     }
 
+    private Specification<User> getLoginSpecificationForSoft(StringFilter loginFilter) {
+        Function<Root<User>, Join<User, UserLogin>> functionToEntity = userRoot -> (Join<User, UserLogin>) userRoot.join(User_.logins);
+        Function<Join<User, UserLogin>, Expression<String>> entityToColumn = entity -> entity.get(UserLogin_.login);
+        return buildSpecificationDistinct(loginFilter, functionToEntity.andThen(entityToColumn));
+    }
+
+    protected Specification<User> buildSpecificationDistinct(StringFilter filter, Function<Root<User>, Expression<String>> metaclassFunction) {
+        if (filter.getEquals() != null) {
+            return equalsSpecificationDistinct(metaclassFunction, filter.getEquals());
+        } else if (filter.getIn() != null) {
+            return valueInDistinct(metaclassFunction, filter.getIn());
+        } else if (filter.getContains() != null) {
+            return likeUpperSpecificationDistinct(metaclassFunction, filter.getContains());
+        } else if (filter.getDoesNotContain() != null) {
+            return doesNotContainSpecificationDistinct(metaclassFunction, filter.getDoesNotContain());
+        } else if (filter.getNotEquals() != null) {
+            return notEqualsSpecificationDistinct(metaclassFunction, filter.getNotEquals());
+        } else if (filter.getSpecified() != null) {
+            return byFieldSpecifiedDistinct(metaclassFunction, filter.getSpecified());
+        }
+        return null;
+    }
+
+    protected <X> Specification<User> equalsSpecificationDistinct(Function<Root<User>, Expression<X>> metaclassFunction, X value) {
+        return (root, query, builder) -> {
+            final Predicate predicate = builder.equal(metaclassFunction.apply(root), value);
+            query.distinct(true);
+            return predicate;
+        };
+    }
+
+    protected <X> Specification<User> notEqualsSpecificationDistinct(Function<Root<User>, Expression<X>> metaclassFunction, X value) {
+        return (root, query, builder) -> {
+            final Predicate predicate = builder.not(builder.equal(metaclassFunction.apply(root), value));
```

Comment:
do we really need `predicate` variable here and similar places? seems we can just return.

Reply:
fixed